### PR TITLE
Add concatenate() and trim() sequences helpers

### PR DESCRIPTION
**Pull request by notwaldorf**
Add a `concatenate()` and `trim()` functions that we used in studio (I haven't found any other ones that were used and we didn't have)

### CODE REVIEW

**Comment by notwaldorf** [comment]
@adarob 👀👀👀

File: music/src/core/sequences.ts [hunk 1]
```diff
@@ -424,3 +424,159 @@ export function mergeInstruments(ns: INoteSequence) {
...
+ */
+export function concatenate(args: INoteSequence[]): NoteSequence {
+  return isQuantizedSequence(args[0]) ?
+      concatenateQuantizedNoteSequence(args) :
+      concatenateUnquantizedNoteSequence(args);
+}
+
+/**
+ * Trim notes from a NoteSequence to lie within a specified time range.
+ * Notes starting before `startTime` are not included. Notes ending after
```
Comment (adarob) [inline music/src/core/sequences.ts]: `start` and `end` (no Time)

File: music/src/core/sequences.ts [hunk 2]
```diff
@@ -424,3 +424,129 @@ export function mergeInstruments(ns: INoteSequence) {
...
+ * `NoteSequences` in `args`. If not provided, the `totalTime` /
+ * `totalQuantizedSteps` of each sequence will be used. Specifying durations is
+ * useful if the sequences to be concatenated are effectively longer than
+ * their `totalTime` (e.g., a sequence that ends with a rest).
+ *
+ * @returns A new sequence that is the result of concatenating the input
+ * sequences.
+ */
+export function concatenate(
+    seqs: INoteSequence[], individualDuration?: number[]): NoteSequence {
```
Comment (adarob) [inline music/src/core/sequences.ts]: use `concatenateSequences` and `sequenceDurations` to mirror py

File: music/src/core/sequences.ts [hunk 3]
```diff
@@ -424,3 +424,129 @@ export function mergeInstruments(ns: INoteSequence) {
...
+  if (individualDuration && individualDuration.length !== seqs.length) {
+    throw new Error(
+        // tslint:disable-next-line:max-line-length
```
Comment (adarob) [inline music/src/core/sequences.ts]: why not just split the string?

File: music/src/core/sequences.ts [hunk 4]
```diff
@@ -424,3 +424,129 @@ export function mergeInstruments(ns: INoteSequence) {
...
+          ns, start, end, 'totalTime', 'startTime', 'endTime',
+          truncateEndNotes);
+}
+
+function concatenateHelper(
+    seqs: INoteSequence[], totalKey: 'totalQuantizedSteps'|'totalTime',
+    startKey: 'startTime'|'quantizedStartStep',
+    endKey: 'endTime'|'quantizedEndStep',
+    individualDuration?: number[]): NoteSequence {
+  let concatSeq: NoteSequence;
```
Comment (adarob) [inline music/src/core/sequences.ts]: you should probably clone the first one to get the tempos, quantinfo, etc

File: music/src/core/sequences.ts [hunk 5]
```diff
@@ -424,3 +424,129 @@ export function mergeInstruments(ns: INoteSequence) {
...
+ * @param individualDuration (Optional) An array of durations for each of the
+ * `NoteSequences` in `args`. If not provided, the `totalTime` /
+ * `totalQuantizedSteps` of each sequence will be used. Specifying durations is
+ * useful if the sequences to be concatenated are effectively longer than
+ * their `totalTime` (e.g., a sequence that ends with a rest).
+ *
+ * @returns A new sequence that is the result of concatenating the input
+ * sequences.
+ */
+export function concatenate(
```
Comment (adarob) [inline music/src/core/sequences.ts]: Can you update the demos to use this?

File: music/src/core/sequences.ts (hunk 4 above)
Comment (notwaldorf) [reply music/src/core/sequences.ts]: I do, see the (i === 0) branch below.

**Pull request merged**